### PR TITLE
[OCPCLOUD-892] Add defaulting/validation for MachineSet ProviderSpecs

### PR DESCRIPTION
**Pull request by JoelSpeed**
This PR introduces defaulting and validation for MachineSet ProviderSpecs.
This should then match the level of defaulting/validation that we have in place for Machines.

TODO:
- [x] Test the webhooks (though Machine doesn't have any either, maybe cover both)

### CODE REVIEW

File: pkg/apis/machine/v1beta1/machineset_webhook.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledMachineSet)
+}
+
+func (h *machineSetValidatorHandler) validateMachineSet(ms *MachineSet) (bool, utilerrors.Aggregate) {
+	var errs []error
+
+	// Create a Machine from the MachineSet and validate the Machine template
+	m := &Machine{Spec: ms.Spec.Template.Spec}
```
Comment (Danil-Grigorev) [inline pkg/apis/machine/v1beta1/machineset_webhook.go:120]: I would add a validator that `selector.matchLabels` is equal to machine labels.

Comment (JoelSpeed) [reply pkg/apis/machine/v1beta1/machineset_webhook.go:120]: We had planned for this PR to just be providing validation for the ProviderSpec. I'm intending to follow up next sprint with a validation for the other fields within MachineSetSpec, this being one of those validations

File: pkg/apis/machine/v1beta1/machineset_webhook.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// machineSetValidatorHandler validates MachineSet API resources.
+// implements type Handler interface.
+// https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/webhook/admission#Handler
+type machineSetValidatorHandler struct {
+	clusterID         string
```
Comment (enxebre) [inline pkg/apis/machine/v1beta1/machineset_webhook.go:18]: since this is now used by 4 types it could be a struct on its own that we embed?

**Comment by enxebre** [comment]
thanks! I dropped a comment, lgtm overall. Needs rebase.

File: pkg/apis/machine/v1beta1/machineset_webhook.go (hunk 2 above)
Comment (JoelSpeed) [reply pkg/apis/machine/v1beta1/machineset_webhook.go:18]: Do you mean the clusterID, webhookOperations and decoder options?

File: pkg/apis/machine/v1beta1/machineset_webhook_test.go [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+			mgr.GetWebhookServer().Register("/mutate-machine-openshift-io-v1beta1-machineset", &webhook.Admission{Handler: machineSetDefaulter})
+			mgr.GetWebhookServer().Register("/validate-machine-openshift-io-v1beta1-machineset", &webhook.Admission{Handler: machineSetValidator})
+
+			ms := &MachineSet{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "machineset-creation-",
+					Namespace:    namespace.Name,
+				},
+				Spec: MachineSetSpec{
+					Template: MachineTemplateSpec{
```
Comment (michaelgugino) [inline pkg/apis/machine/v1beta1/machineset_webhook_test.go:170]: Looking at our API, Template is actually an optional field.  We should ensure that we handle that case and we don't crash.

Comment (michaelgugino) [reply pkg/apis/machine/v1beta1/machineset_webhook_test.go:170]: Also, Template.Spec is also optional.

File: pkg/apis/machine/v1beta1/machineset_webhook_test.go [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+			platformType: osconfigv1.GCPPlatformType,
+			clusterID:    "gcp-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &gcp.GCPMachineProviderSpec{
+					Region: "region",
+					Zone:   "region-zone",
+				},
+			},
+			expectedError: "",
+		},
```
Comment (michaelgugino) [inline pkg/apis/machine/v1beta1/machineset_webhook_test.go:126]: ProviderSpec.Value can be nil.

```
		{
			name:         "with GCP and nil providerSpec",
			platformType: osconfigv1.GCPPlatformType,
			clusterID:    "gcp-cluster",
			providerSpecValue: nil,
			expectedError: "x",
		},
```

File: pkg/apis/machine/v1beta1/machineset_webhook.go (hunk 2 above)
Comment (enxebre) [reply pkg/apis/machine/v1beta1/machineset_webhook.go:18]: yep, wdyt?

Comment (JoelSpeed) [reply pkg/apis/machine/v1beta1/machineset_webhook.go:18]: Sounds good, I'll take a look

**Comment by michaelgugino** [comment]
I'm seeing some crazy stuff locally after adding the modifications from my PR: https://gist.github.com/michaelgugino/8efdf148902b84cf1ea03c71d4c85b04

On one run of the tests, I'm getting create errors for "TestMachineSetUpdate/with_a_valid_Azure_ProviderSpec" and it's complaining about unable to marshall JSON for AWS.

Then on the next run, segfault.

Just seems to affect TestUpdate.  If I remove that test, my test and the other tests work.

File: pkg/apis/machine/v1beta1/machineset_webhook_test.go (hunk 3 above)
Comment (JoelSpeed) [reply pkg/apis/machine/v1beta1/machineset_webhook_test.go:170]: While both fields are marked as `omitempty`, as neither is a pointer value, the go type system will infer them even if they are empty, so we don't need to do any checks for nils or anything, I don't think there is anything to handle specifically for these, unless I have missed something?

File: pkg/apis/machine/v1beta1/machineset_webhook_test.go (hunk 4 above)
Comment (JoelSpeed) [reply pkg/apis/machine/v1beta1/machineset_webhook_test.go:126]: Good catch, added some tests and a check that ensures the provider spec value isn't nil

**Comment by enxebre** [comment]
Nice!
/approve
/hold
@JoelSpeed to manually verify the e2e test results before unholding as some tests that might be broken by this PR are not blocking in this repo, e.g e2e-azure-operator, e2e-azure.

**Comment by JoelSpeed** [comment]
/hold cancel

I've manually verified that you can bring up an IPI cluster from a release made from this branch and it functions as expected on Azure and AWS, so I'm confident that this shouldn't be breaking our E2E suite even if we can't get Azure to pass

**Pull request merged**